### ISSUE TITLE
Add type hinting in the codebase

### ISSUE THREAD

**Issue by ankurankan** [pgmpy/pgmpy#1270]
From python 3.4, type hinting feature has been added to specify the expected type of arguments in functions/methods/classes and their returns types. Modify our codebase to have type hintings as it makes the code much more readable. 

Most user facing classes are in the following files:
- [ ] DAG.py
- [ ] DiscreteBayesianNetwork.py
- [ ] LinearGaussianBayesianNetwork.py
- [ ] FunctionalBayesianNetwork.py
- [ ] PC.py
- [ ] HillClimbSearch.py
- [ ] GES.py
- [ ] expert.py
- [ ] MLE.py
- [ ] EM.py
- [ ] BayesianEstimator.py

**Comment by kyle-pena-nlp** [comment]
Are you also interested in type checking, a la typeguard?: https://pypi.org/project/typeguard/

**Comment by ankurankan** [comment]
@kyle-pena-nlp I had a quick look at typeguard but I don't think I would like that to be added yet. I expect some of our object types to be a bit complicated and I am not yet sure how easy would it be to define them using the `typing` module. So, as a first step, I would instead just like to see how well just the type hinting performs. And if it works reasonably well, we can add typeguard too.

**Comment by DevAnuragT** [comment]
Hi! I'm interested in contributing to this issue. Since it's still open and unassigned, could I take it up?

**Comment by ankurankan** [comment]
@DevAnuragT Sure, I would also suggest starting with some of the more user facing classes such as `pgmpy.base.DAG`, `pgmpy.base.PDAG`, `pgmpy.models.*BayesianNetowrk`.

**Comment by arocapedro** [comment]
I have some work put in this, but I never got around pushing my changes. I will do it tomorrow!

**Comment by ankurankan** [comment]
@arocapedro Yes, I completely forgot about it. I will add a list of files here, so that we can keep a track of which files still have type hinting missing.

**Comment by DevAnuragT** [comment]
I’ve started working on this issue and will be raising separate PRs for each file to keep the changes focused and easier to review.

First PR: #2077  — Adds type hints to `DAG.py`.
Second PR: #2081 — Adds type hints to `DiscreteBayesianNetwork.py`.
Third PR: #2082 — Adds type hints to `LinearGaussianBayesianNetwork.py`.
Fourth PR: #2083 — Adds type hints to `FunctionalBayesianNetwork.py`.

Will continue with the rest of the files in subsequent PRs.

**Comment by Hridayy10** [comment]
Hi, I’d like to work on this issue as part of the European Summer of Code. Could you please assign it to me?

**Comment by Jatinbhardwaj-093** [comment]
Hello @ankurankan,

I would like to work on this issue, as I have good experience with type handling in Python. I can update most of the repository with basic type annotations for pgmpy classes and built-in Python types such as list, dict, str, etc in one single PR.

Since Python 3.10+, we can also use modern type hinting (for example, List → list, Union → |). I would like to help update the current codebase to follow these newer typing conventions.

We could also consider developing something similar to `numpy.typing` for better typing support in pgmpy. I plan to open a separate PR for a more systematic typing approach in pgmpy.

For now, could you please assign this issue to me so I can start with a PR adding the basic type hints?
